src/metric/auc.cu

```diff
@@ -46,7 +45,7 @@ struct DeviceAUCCache {
   dh::device_vector<size_t> unique_idx;
   // p^T: transposed prediction matrix, used by MultiClassAUC
   dh::device_vector<float> predts_t;
-  std::unique_ptr<dh::AllReducer> reducer;
+  collective::DeviceCommunicator* communicator;
```

Comment:
If this is now a global instance, we don't have to maintain a pointer to it.

Reply:
We still need the device id to get the communicator, which in this class is only passed in during `Init` and not saved. We can probably clean this up once we have better device id management.

Reply:
Ended up removing these pointers.